### DISCUSSION TITLE
Multiple sesion en esp32 via wifi

### DISCUSSION THREAD

**Discussion by Veroledez** [lathoub/Arduino-AppleMIDI-Library#133] [Q&A] [answered]
Hola, gracias por tan fabulosa libreria.
Bien, sere breve, me preguntaba si se pueden crear varias sesiones en el es32 via wifi, asi como se crean via ethernet

APPLEMIDI_CREATE_INSTANCE (EthernetUDP, MIDI1, " Arduino1 " , DEFAULT_CONTROL_PORT);
APPLEMIDI_CREATE_INSTANCE (EthernetUDP, MIDI2, " Arduino2 " , DEFAULT_CONTROL_PORT + 2 );

Pero no se si se pueda, y que habria que colocar en lugar de EthernetUDP.

Muchas gracias, un saludo

**Comment by lathoub** [comment]
Hola @Veroledez 

Taken from the [ESP32_wifi example](https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/master/examples/ESP32_NoteOnOffEverySec/ESP32_NoteOnOffEverySec.ino) and 

https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/7c96535298b2cbc0d42cfadd4d5cc4188a69a277/src/AppleMIDI.h#L350-L352

Your code could something like this:

```cpp
#include <WiFi.h>
#include <WiFiClient.h>
#include <WiFiUdp.h>

APPLEMIDI_CREATE_INSTANCE(WiFiUDP, MIDI1, "AppleMIDI-ESP32-1", DEFAULT_CONTROL_PORT);
APPLEMIDI_CREATE_INSTANCE(WiFiUDP, MIDI2, "AppleMIDI-ESP32-2", DEFAULT_CONTROL_PORT+2);

```

Let me know if this works

**Reply by Veroledez** [reply]
Hola, lo he probado y en el rtp  midi no se anuncia el esp32, cuando uso solo el   ejemplo de una instancia si se anuncia

**Reply by lathoub** [reply]
> Hola, lo he probado y en el rtp midi no se anuncia el esp32, cuando uso solo el ejemplo de una instancia si se anuncia

Can you post minimalist code, so i can try to replicate

**Reply by Veroledez** [reply]
Hola, aqui  esta mi codigo completo, soy un novato en esto por lo que algunas lineas  que tiraban error al compilar y como no se que poner alli las he comentado, y aunque compila y sube al esp32 no hay forma de conectarse a la sesion. Lo ejecuto en windows 10, con rtpmidi y fl studio, el ejemplo predeterminado de la libreria si me funciona
Agradesco mucho tu ayuda, un saludo
`  
#include <WiFi.h>
#include <WiFiClient.h>
#include <WiFiUdp.h>

#define NO_SESSION_NAME
#define SerialMon Serial
#define APPLEMIDI_DEBUG SerialMon
#include <AppleMIDI.h>

char ssid[] = "HOMERED"; //  your network SSID (name)
char pass[] = "12345678";    // your network password (use for WPA, or use as key for WEP)

unsigned long t0 = millis();
unsigned long t1 = millis();
int8_t isConnected = 0;

APPLEMIDI_CREATE_INSTANCE(WiFiUDP, MIDI1, " ESP-1", DEFAULT_CONTROL_PORT);
APPLEMIDI_CREATE_INSTANCE(WiFiUDP, MIDI2, " ESP-2", DEFAULT_CONTROL_PORT + 2);
// -----------------------------------------------------------------------------
//
// -----------------------------------------------------------------------------
void setup()
{
  DBG_SETUP(115200);
  DBG("Booting");

  WiFi.begin(ssid, pass);

  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    DBG("Establishing connection to WiFi..");
  }
  DBG("Connected to network");

  DBG(F("OK, now make sure you an rtpMIDI session that is Enabled"));
//  DBG(F("Add device named Arduino with Host"), WiFiUDP.localIP(), "Port", MIDI1.getPort(), "(Name", MIDI1.getName(), ")");
  DBG(F("Select and then press the Connect button"));
  DBG(F("Then open a MIDI listener and monitor incoming notes"));
  DBG(F("Listen to incoming MIDI commands"));

  MIDI1.begin();
  MIDI2.begin();
//  AppleMIDI.setHandleConnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc, const char* name) {
    isConnected++;
//    DBG(F("Connected to session"), ssrc, name);
  //});
 // AppleMIDI.setHandleDisconnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc) {
    isConnected--;
 //   DBG(F("Disconnected"), ssrc);
 // });
  
  MIDI1.setHandleNoteOn([](byte channel, byte note, byte velocity) {
    DBG(F("NoteOn"), note);
  });
  MIDI1.setHandleNoteOff([](byte channel, byte note, byte velocity) {
    DBG(F("NoteOff"), note);
  });

  DBG(F("Sending NoteOn/Off of note 45, every second"));
}

// -----------------------------------------------------------------------------
//
// -----------------------------------------------------------------------------
void loop()
{
  // Listen to incoming notes
  MIDI1.read();
MIDI2.read();
  // send a note every second
  // (dont cáll delay(1000) as it will stall the pipeline)
  if ( (millis() - t0) > 1000)
  {
    t0 = millis();

    byte note = 40;
    byte velocity = 64;
    byte channel = 1;

    MIDI1.sendNoteOn(note, velocity, channel);
    MIDI1.sendNoteOff(note, velocity, channel);

  }


    if ( (millis() - t1) > 1300)
  {
    t1 = millis();

    byte note = 50;
    byte velocity = 127;
    byte channel = 1;


    MIDI2.sendNoteOn(note, velocity, channel);
    MIDI2.sendNoteOff(note, velocity, channel);
  }
}


  `

**Comment by lathoub** [comment, marked as answer]
Hola @Veroledez. At first glace nothing wrong with your code, good job!

I copied into the Arduino IDE and compiled for ESP32 DevKit, tweaked the code left and right (kept the essentials) and got good results. Can you try the below.

I noticed you did not show the IP address - how did you know the IP address of the ESP32? (Bonjour is **not** enabled in the code below)

In any case, try the below:

```cpp
#include <WiFi.h>
#include <WiFiClient.h>
#include <WiFiUdp.h>

#define SerialMon Serial
#define APPLEMIDI_DEBUG SerialMon
#include <AppleMIDI.h>

char ssid[] = "xxx"; // your network SSID (name)
char pass[] = "yyy"; // your network password (use for WPA, or use as key for WEP)

unsigned long t0 = millis();
int8_t isConnected = 0;

APPLEMIDI_CREATE_INSTANCE(WiFiUDP, MIDI1, " ESP-1", DEFAULT_CONTROL_PORT);
APPLEMIDI_CREATE_INSTANCE(WiFiUDP, MIDI2, " ESP-2", DEFAULT_CONTROL_PORT + 2);

// -----------------------------------------------------------------------------
//
// -----------------------------------------------------------------------------
void setup()
{
  DBG_SETUP(115200);
  DBG("Booting");

  WiFi.begin(ssid, pass);

  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    DBG("Establishing connection to WiFi..");
  }
  DBG("Connected to network");

  DBG(F("OK, now make sure you an rtpMIDI session that is Enabled"));
  DBG(F("Add device named Arduino with Host"), WiFi.localIP(), "Port", AppleMIDI1.getPort(), "and Port", AppleMIDI2.getPort());
  DBG(F("Select and then press the Connect button"));
  DBG(F("Then open a MIDI listener and monitor incoming notes"));
  DBG(F("Listen to incoming MIDI commands"));

  AppleMIDI1.setHandleConnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc, const char* name) {
    isConnected++;
    DBG(F("MIDI 1 Connected to session"), ssrc, name);
  });
  AppleMIDI1.setHandleDisconnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc) {
    isConnected--;
    DBG(F("MIDI 1 Disconnected"), ssrc);
  });

  AppleMIDI2.setHandleConnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc, const char* name) {
    isConnected++;
    DBG(F("MIDI 2 Connected to session"), ssrc, name);
  });
  AppleMIDI2.setHandleDisconnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc) {
    isConnected--;
    DBG(F("MIDI 2 Disconnected"), ssrc);
  });

  MIDI1.begin();
  MIDI2.begin();

  MIDI1.setHandleNoteOn([](byte channel, byte note, byte velocity) {
    DBG(F("MIDI 1 NoteOn"), note);
  });
  MIDI2.setHandleNoteOn([](byte channel, byte note, byte velocity) {
    DBG(F("MIDI 2 NoteOff"), note);
  });

  DBG(F("Sending NoteOn/Off of note 45, every second"));
}

// -----------------------------------------------------------------------------
//
// -----------------------------------------------------------------------------
void loop()
{
  // Listen to incoming notes
  MIDI1.read();
  MIDI2.read();

  // send a note every second
  // (dont cáll delay(1000) as it will stall the pipeline)
  if (isConnected >= 2 && (millis() - t0) > 1000)
  {
    t0 = millis();

    MIDI1.sendNoteOn(40, 64, 1);
    MIDI2.sendNoteOn(40, 64, 1);
  }
}
```

**Reply by Veroledez** [reply]
Hola, ya he tenido tiempo de probarlo y funciona muy bien, muchas gracias
